File: src/test/java/com/meti/ServerTest.java
package com.meti;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.net.URL;

import static org.junit.jupiter.api.Assertions.assertEquals;

class ServerTest {
    private Server server;

    @Test
    void construct() throws IOException {
        URL url = new URL("http://localhost:80");
        byte[] bytes = URLUtils.readAllBytes(url);
        assertEquals("test", new String(bytes));
    }

    @AfterEach
    void tearDown() {
        server.stop();
    }

    @BeforeEach
    void setUp() throws IOException {
        server = new NanoServer(80, new SingletonRouter(new TestRoute()));
        server.start();
    }

    static class TestRoute implements Route {
        @Override
        public ByteResponse process() {
            return new ByteResponse(DefaultCode.OK, DefaultType.PLAIN, "test".getBytes());
        }

        @Override
        public boolean canProcess(Request request) {
            return true;
        }
    }
}

File: src/test/scala/com/meti/ServerTest.scala
package com.meti

import java.net.URL

import com.meti.response.{DefaultCode, PlainType, Response, StringResponse}
import com.meti.route.{Route, SingletonRouter}
import org.junit.jupiter.api.Assertions._
import org.junit.jupiter.api.{AfterEach, BeforeEach, Test}

class ServerTest {
	var server: Option[Server] = None

	@BeforeEach
	def setUp(): Unit = {
		val router = SingletonRouter(TestRoute)
		server = Some(NanoServer(80, router))
		server.get.start()
	}

	@AfterEach
	def tearDown(): Unit = {
		server.get.stop()
	}

	@Test
	def construct(): Unit = {
		val url = new URL("http://localhost:80")
		val stream = url.openStream()
		val data = stream.readAllBytes()
		assertEquals("test", new String(data))
		stream.close()
	}

	object TestRoute extends Route {
		override def process(request: Request): Response = StringResponse(DefaultCode.OK, PlainType, "test")
	}

}
